[SofaKernel] Replacing const char* with strings for group / help / widget etc.

marques-bruno opened the pull request:
In BaseData and BaseLink mainly.



______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

marques-bruno committed:
Replacing const char* with strings for group / help / widget etc. in Data & Links

epernod commented on applications/sofa/gui/qt/QDisplayPropertyWidget.cpp:
```diff
@@ -161,7 +161,7 @@ void QDisplayPropertyWidget::addComponent(const QString& component, core::object
 
         // for each data of the current object we determine where it belongs
-        QString group = data->getGroup();
+        QString group = data->getGroup().c_str();
```
normally there is a method from Qt to convert directly std::string to QString. Maybe better than recasting the string to const char* before giving it ton QString.
(probably it does the same internally, but it is less dirty)

marques-bruno replied:
Makes sense, I'll check for that Qt method

epernod commented on SofaKernel/framework/sofa/core/objectmodel/BaseData.h:
```diff
@@ -78,13 +78,13 @@ class SOFA_CORE_API BaseData : public DDGNode
+        std::string name;
+        std::string ownerClass;
+        std::string group;
+        std::string widget;
```
are all those info of the Data only used for gui?

marques-bruno replied:
> Error from CI :
> 
> ```
> Binding_Data.cpp:866:35: error: cannot convert ‘const string {aka const std::__cxx11::basic_string<char>}’ to ‘const char*’ in initialization
>      const char *h = data->getHelp() ;
> ```
> 
> Could you also give explanation for this change @marques-bruno ?

It's a breaking change and I don't compile with the same settings as the CI. I'll fix those issues as they show up.
Regarding the "why":

I had some segmentation faults with strings that were badly set in BaseData::group and since char* is so 80's I thought it's the occasion to replace them with std::strings, like a normal c++ developer does :)
It solved my problem, hence the PR

marques-bruno replied:
> are all those info of the Data only used for gui?

name, help and group are intensively used in SofaPython and SofaPython3. OwnerClass could also be. widget is probably exclusively used by the GUI

guparan replied:
This change will also fix #1092 :+1:

guparan replied:
:warning: Also, as stated in #1092, don't forget to revert #1090 (temporary fix).

marques-bruno replied:
> warning Also, as stated in #1092, don't forget to revert #1090 (temporary fix).

The temp fix in #1090 is overridden by the PR already, so nothing to revert :+1:

epernod commented on SofaKernel/framework/sofa/core/objectmodel/BaseData.h:
```diff
@@ -157,25 +157,25 @@ class SOFA_CORE_API BaseData : public DDGNode
 
     /// Get owner class
-    const char* getOwnerClass() const { return ownerClass; }
+    const std::string& getOwnerClass() const { return ownerClass; }
```
maybe for backport it would be good to keep some version returing c_char . Where you do the conversion.

marques-bruno replied:
I think that would only delay the moment these things will have to be fixed. I can leave a method with a [[deprecated]] tag maybe, but I would still have to change the method name (`getGroup_as_cstring()` for instance), as polymorphism over return types is not allowed in c++, so the code would still be breaking

epernod replied:
ok I see. Then replace to string.
I will do the conversion on my side if needed ;)

marques-bruno committed:
Apply changes to SofaPython

epernod commented:
[ci-build][with-all-tests]

marques-bruno committed:
Replacing QString(XXX.c_str()) with QString::fromStdString(XXX)

marques-bruno committed:
Replacing QString(XXX.c_str()) with QString::fromStdString(XXX)

marques-bruno committed:
Reverting fromStdString for QLinkSimpleEdit

Which seems to be the only Qt function taking a char* instead of a QString...

marques-bruno committed:
...

marques-bruno committed:
...

damienmarchal commented on SofaKernel/framework/sofa/core/objectmodel/BaseData.h:
```diff
@@ -319,13 +319,13 @@ class SOFA_CORE_API BaseData : public DDGNode
 
     /// Help message
-    const char* help {nullptr} ;
+    std::string help {nullptr} ;
```
I'm not sure this is ok.

EDIT: to be more assertive...I'm sure this is not ok.

damienmarchal commented:
Hi all, I think this PR breaks a lot of things. 
Don't merge it unless clarified if all the subclass of Data<> (eg: DataFileName, BaseTopologyData)
that are having an helpMsg == nullptr are not crashing Sofa.

damienmarchal committed:
[SofaKernel] Fix conversion to std::string.

Simply replacing char* with std::string is dangerous as every caller's points will
that was doing asigning NULL/nullptr will become undefined behavior.

The solution implemented here is to keep the old char* signature to enforce
proper null coercion to empty string.

damienmarchal committed:
[SofaKernel] Add deprecation message and doc

marques-bruno committed:
Merge pull request #1 from SofaDefrost/charptr_to_str

Charptr to str

marques-bruno committed:
FIX: ownerClass issue in runSofa's tooltips

damienmarchal committed:
Update BaseData.h

damienmarchal committed:
Update Data.h

marques-bruno commented:
@damienmarchal, this PR be set back to to_review with your fix right?

damienmarchal commented:
Yes.

damienmarchal committed:
Update BaseData.h

marques-bruno commented:
[ci-build][with-all-tests]

fredroy commented:
[ci-build][force-full-build]

marques-bruno committed:
Merge branch 'master' into charptr_to_string

fredroy commented:
[ci-build][with-all-tests]
